04_Scheduling/04_Locks.md

```diff
@@ -0,0 +1,205 @@
+# Critical Sections and locks
+
+## Introduction
+
+Now that we have introduced a multi tasking mechanism our kernel is able to run different programs concurrently. But now this can introduce new problems, having concurrent process being able to access same resources at nearly the "same time". Let's start with an example scenario.
+
+Imagine that we have a shared resource that can be accessed at a specified address (that resource can be anything: an i/o memory mapped device, a memory mapped hard drive location, a shared variable, or whatever).
+
+Let's assume for example that the shared resource is accessible at the address 0xDEADBEEF, this resource is a serial port connected to an output device. 
+
+Now we have initially a single process that is sending a string to the serial port one character at time: 
+
+```c 
+#define SHARED_RESOURCE 0xDEADBEEF 
+
+// Other code doing other stuff
+char string_to_send[] = "I am the first string"
+
+int i = 0;
+while(i < strlen(string_to_send) {
+    *((int *) shared_resource) = string_to_send[i++];
```

Comment:
It'd probably make more sense for this to be a `char*` I think :P

Reply:
yeah sorry leftover from previous example XD